File: python/transparency/alpha_transparency/alpha_transparency.py
import sys, os, ctypes
import os, sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

# alpha_transparency -- Python (ctypes) port of the VB6 mirror.
# Draws an image at fill alpha 0.5 and a second at the default alpha 1.0
# using extended graphics states.

HERE = os.path.dirname(os.path.abspath(__file__))
IMG = os.path.join(HERE, "..", "..", "..", "test_files", "images", "tree-frog-69813_640.jpg")

clWhite = 0xFFFFFF
clBlack = 0x000000


def _err(data, code, msg, typ):
    if msg:
        print(msg.decode('latin-1', 'replace'))
    return 0


_cb = L.TErrorProc(_err)  # keep global ref


def main():
    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, _cb)
    L.pdfCreateNewPDFA(pdf, b"")

    L.pdfSetPageCoords(pdf, L.pcTopDown)
    L.pdfSetUseTransparency(pdf, 0)  # disable color key masking for images

    L.pdfAppend(pdf)

    L.pdfSetFontA(pdf, b"Helvetica", L.fsRegular, 12.0, 0, L.cp1252)
    L.pdfWriteTextA(pdf, 50.0, 50.0, b"Fill Alpha = 0.5")

    L.pdfRectangle(pdf, 50.0, 70.0, 110.0, 160.0, L.fmFill)
    L.pdfSetFillColor(pdf, clWhite)
    L.pdfWriteTextA(pdf, 55.0, 75.0, b"Background")

    g = L.TPDFExtGState()
    L.pdfInitExtGState(ctypes.byref(g))
    g.FillAlpha = 0.5
    gs = L.pdfCreateExtGState(pdf, ctypes.byref(g))
    L.pdfSetExtGState(pdf, gs)

    img = L.pdfInsertImageExA(pdf, 60.0, 84.0, 200.0, 0.0, IMG.encode('latin-1'), 0)

    # Restore by creating a second state that undoes the change and activate it.
    g.FillAlpha = 1.0
    gs = L.pdfCreateExtGState(pdf, ctypes.byref(g))
    L.pdfSetExtGState(pdf, gs)

    L.pdfSetFillColor(pdf, clBlack)
    L.pdfWriteTextA(pdf, 340.0, 50.0, b"Fill Alpha = 1.0 (default)")
    L.pdfRectangle(pdf, 340.0, 70.0, 110.0, 160.0, L.fmFill)
    L.pdfSetFillColor(pdf, clWhite)
    L.pdfWriteTextA(pdf, 345.0, 75.0, b"Background")
    L.pdfPlaceImage(pdf, img, 350.0, 84.0, 200.0, 0.0)

    L.pdfEndPage(pdf)

    if L.pdfHaveOpenDoc(pdf) != 0:
        outFile = os.path.join(HERE, "out.pdf")
        if L.pdfOpenOutputFileA(pdf, outFile.encode('latin-1')) == 0:
            L.pdfDeletePDF(pdf)
            return
        if L.pdfCloseFile(pdf) != 0:
            print('PDF file "' + outFile + '" successfully created!')

    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()

File: cpp-linux/transparency/alpha_transparency/alpha_transparency.cpp
// alpha_transparency -- C++ mirror of
// examples\Vb6\transparency\alpha_transparency\alpha_transparency.bas
// Draws an image at fill alpha 0.5 and a second at the default alpha 1.0 using
// extended graphics states.
#include "apputil.h"

static const unsigned long clWhite = 0xFFFFFF;
static const unsigned long clBlack = 0x0;

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType) {
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

int main() {
    ChdirToExe();
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, PDFError);
    pdfCreateNewPDFA(pdf, "");
    pdfSetPageCoords(pdf, pcTopDown);
    pdfSetUseTransparency(pdf, 0); // Disable color key masking for images

    pdfAppend(pdf);
        pdfSetFontA(pdf, "Helvetica", fsRegular, 12.0, 0, cp1252);
        pdfWriteTextA(pdf, 50, 50, "Fill Alpha = 0.5");

        pdfRectangle(pdf, 50, 70, 110, 160, fmFill);
        pdfSetFillColor(pdf, clWhite);
        pdfWriteTextA(pdf, 55, 75, "Background");

        TPDFExtGState g;
        pdfInitExtGState(&g);
        g.FillAlpha = 0.5;
        int gs = pdfCreateExtGState(pdf, &g);
        pdfSetExtGState(pdf, gs);

        int img = pdfInsertImageExA(pdf, 60, 84, 200, 0, LUMAS_REPO_ROOT "/images/photo_02.jpg", 0);

        // Restore by creating a second state that reverts the change.
        g.FillAlpha = 1.0;
        gs = pdfCreateExtGState(pdf, &g);
        pdfSetExtGState(pdf, gs);

        pdfSetFillColor(pdf, clBlack);
        pdfWriteTextA(pdf, 340, 50, "Fill Alpha = 1.0 (default)");
        pdfRectangle(pdf, 340, 70, 110, 160, fmFill);
        pdfSetFillColor(pdf, clWhite);
        pdfWriteTextA(pdf, 345, 75, "Background");
        pdfPlaceImage(pdf, img, 350, 84, 200, 0);
    pdfEndPage(pdf);

    if (pdfHaveOpenDoc(pdf) != 0) {
        const char* outFile = "out.pdf";
        if (pdfOpenOutputFileA(pdf, outFile) == 0) { pdfDeletePDF(pdf); return 0; }
        if (pdfCloseFile(pdf) != 0)
            printf("PDF file \"%s\" successfully created!\n", outFile);
    }

    pdfDeletePDF(pdf);
    return 0;
}
